src/librbd/io/QosImageDispatch.cc

```diff
@@ -304,9 +310,14 @@ bool QosImageDispatch<I>::needs_throttle(
                      << "flag=" << flag << dendl;
       all_qos_flags_set = false;
     } else {
-      all_qos_flags_set = set_throttle_flag(image_dispatch_flags, flag);
+      flags_to_set |= flag;
     }
   }
+
+  if (all_qos_flags_set) {
```

Comment:
This is wrong, should be `if (flags_to_set != 0) {`

Example:

all_qos_flags_set = true; // initial state
for loop sets some flags in flags_to_set
for loop starts at least one throttle and sets all_qos_flags_set to false
this if fails even though there are flags in flags_to_set